include/laser_filters/median_spatial_filter.h

```diff
@@ -0,0 +1,122 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
+*
+*  Copyright (c) 2017, laser_filters authors
+*  All rights reserved.
+*
+*  Redistribution and use in source and binary forms, with or without
+*  modification, are permitted provided that the following conditions
+*  are met:
+*
+*   * Redistributions of source code must retain the above copyright
+*     notice, this list of conditions and the following disclaimer.
+*   * Redistributions in binary form must reproduce the above
+*     copyright notice, this list of conditions and the following
+*     disclaimer in the documentation and/or other materials provided
+*     with the distribution.
+*   * Neither the name of the Willow Garage nor the names of its
+*     contributors may be used to endorse or promote products derived
+*     from this software without specific prior written permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+/*
+\author Yannic Bachmann
+*/
+
+#ifndef LASER_SCAN_MEDIAN_SPATIAL_FILTER_H
+#define LASER_SCAN_MEDIAN_SPATIAL_FILTER_H
+
+#include "filters/filter_base.hpp"
+#include <sensor_msgs/msg/laser_scan.hpp>
+#include <algorithm>
+#include <vector>
+#include <stdexcept>
+
+namespace laser_filters
+{
+
+class LaserScanMedianSpatialFilter : public filters::FilterBase<sensor_msgs::msg::LaserScan>
+{
+public:
+  int window_size_;
+
+  bool configure()
+  {
+    // Default window size
+    window_size_ = 3;
+    getParam("window_size", window_size_);
+
+    // Ensure window size is positive
+    if (window_size_ <= 0)
+    {
+      throw std::runtime_error("Window size must be positive");
+    }
+
+    // Ensure window size is odd
+    if (window_size_ % 2 == 0)
+    {
+      window_size_ += 1;
```

Comment:
Is it necessary to modify the window size here? Typically the median of a set with an even number of elements is equal to the average of the center two elements - people might find that behavior more expected.

Reply:
The reasoning behind this is that we always modify the current range (`filtered_scan.ranges[i]`) by looking at the ranges around it (e.g. the one directly left and directly right of it if `window_size_=3`). If we have an even window_size, e.g. 4, we would need to either look at two values to the right and one value to the left or the other way around. In other words: If the window_size is even, the current range, which will be modified cannot be centered in the window. 
I would propose to keep it this way but add a warning output if the window_size is even and we have to increment it.

Reply:
Yes, your explanation makes sense. Sounds good to me.